### PR TITLE
Add nn module system

### PR DESCRIPTION
**Pull request by Frikallo**
# Add nn module system for inference

## Summary

Adds a PyTorch-style `nn::Module` system for structured inference. Provides a base class with hierarchical parameter ownership, `load_state_dict()` from HuggingFace safetensors, recursive `to(Device)`, and standard layers wrapping existing `ops::` functions.

## Motivation

Axiom has mature functional ops and safetensors loading, but building any model (Whisper, LLaMA) requires manually wiring dozens of weight tensors by name. This PR adds the thin module abstraction needed to define models as composable structs with automatic weight loading.

## Design

- **Inference-only** — no autograd, no training-specific state
- **Pointer-based parameter registration** — zero overhead in forward passes
- **Non-copyable/non-movable modules** — prevents dangling pointer bugs
- **Configuration-only constructors** — dimensions come from loaded weights, not constructor args
- **Wraps existing ops** — every layer delegates to `ops::matmul`, `ops::conv1d`, `ops::layer_norm`, etc.

### Base class (`nn::Module`)
- `register_parameter(name, tensor)` / `register_module(name, submodule)` — pointer-based registration
- `load_state_dict(map, prefix, strict)` — hierarchical name resolution matching HuggingFace conventions
- `to(Device)` — moves all parameters recursively, skips uninitialized tensors
- `named_parameters(prefix)` / `parameters()` — recursive introspection

### Standard layers
| Layer | Parameters | Wraps |
|-------|-----------|-------|
| `Linear` | weight, bias (optional) | `ops::matmul` + `ops::add` |
| `Embedding` | weight | `ops::embedding` |
| `LayerNorm` | weight, bias | `ops::layer_norm` |
| `RMSNorm` | weight | `ops::rms_norm` |
| `Conv1d` | weight, bias (optional) | `ops::conv1d` |
| `Conv2d` | weight, bias (optional) | `ops::conv2d` |
| `MultiHeadAttention` | q/k/v/out projections (4x Linear) | scaled dot-product attention |
| `ModuleList` | owns `unique_ptr<Module>` children | typed `emplace_back<T>()` |

### Safety
- All `forward()` methods throw `RuntimeError` if called before `load_state_dict()`
- `MultiHeadAttention::forward()` validates 3D input shape and `d_model % num_heads == 0`
- `ModuleList` bounds-checks indexing
- `load_state_dict` throws `ValueError` on missing keys when `strict=true`

## File structure

```
include/axiom/nn.hpp                    # Umbrella include
include/axiom/nn/module.hpp             # Base class
include/axiom/nn/linear.hpp
include/axiom/nn/embedding.hpp
include/axiom/nn/normalization.hpp      # LayerNorm, RMSNorm
include/axiom/nn/conv.hpp              # Conv1d, Conv2d
include/axiom/nn/container.hpp         # ModuleList
include/axiom/nn/attention.hpp         # MultiHeadAttention

src/nn/module.cpp
src/nn/linear.cpp
src/nn/embedding.cpp
src/nn/normalization.cpp
src/nn/conv.cpp
src/nn/container.cpp
src/nn/attention.cpp

tests/test_nn_module.cpp               # 13 tests: registration, state_dict, to()
tests/test_nn_layers.cpp               # 27 tests: all layers + uninitialized checks
tests/test_nn_attention.cpp            # 9 tests: structure, forward, masking, validation
```

## Tests

49 new tests covering:
- Parameter registration and hierarchical name resolution
- `load_state_dict` with prefix, strict/non-strict, extra keys, missing keys
- `to(Device)` with initialized and uninitialized parameters
- Forward correctness for every layer (values verified against manual computation)
- Batched forward for Linear (3D input)
- Composite modules (two-layer MLP end-to-end)
- `ModuleList` with `emplace_back`, indexing, recursive parameter collection
- `MultiHeadAttention` with identity projections, causal masking, prefix loading
- Error paths: forward before load, invalid input dimensions, out-of-bounds indexing
- Integration: `ModuleList` of transformer layers with deep parameter hierarchy

All 1075 tests pass (49 new + 1026 existing unchanged).

## Usage example

```cpp
#include <axiom/axiom.hpp>
#include <axiom/nn.hpp>

struct WhisperEncoder : axiom::nn::Module {
    axiom::nn::Conv1d conv1_{1, 1};  // stride=1, padding=1
    axiom::nn::Conv1d conv2_{2, 1};  // stride=2, padding=1
    axiom::nn::ModuleList layers_;
    axiom::nn::LayerNorm ln_post_;

    WhisperEncoder(int n_layers, int n_heads) {
        register_module("conv1", conv1_);
        register_module("conv2", conv2_);
        register_module("layers", layers_);
        register_module("ln_post", ln_post_);
        for (int i = 0; i < n_layers; ++i)
            layers_.emplace_back<WhisperEncoderLayer>(n_heads);
    }

    axiom::Tensor forward(const axiom::Tensor& mel) const {
        auto x = conv1_(mel).gelu();
        x = conv2_(x).gelu();
        for (size_t i = 0; i < layers_.size(); ++i) {
            auto& layer = static_cast<const WhisperEncoderLayer&>(layers_[i]);
            x = layer.forward(x);
        }
        return ln_post_(x);
    }
};

// Load and run
auto weights = axiom::io::safetensors::load("model.safetensors");
WhisperEncoder encoder(4, 6);
encoder.load_state_dict(weights, "model.encoder");
encoder.to(axiom::Device::GPU);
auto output = encoder.forward(mel_spectrogram);
```

## Known limitations

- `load_state_dict` silently ignores extra keys in the state dict (matches non-strict PyTorch behavior even when `strict=true` for extra keys — only missing keys are checked)
- `ModuleList` iteration yields `unique_ptr<Module>&`, requiring `(*it)->forward()` syntax
- `masked_fill` on non-contiguous 4D tensors from transposed matmul requires explicit `ascontiguousarray()` (worked around in attention implementation)

### CODE REVIEW

**Comment by Frikallo** [comment]
+ Updated docs and included a Whisper POC to prove Axiom's readiness for SafeTensor inference in a PyTorch-friendly way!
